FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+transitions from going forward (doing) to going backwards (undoing) as the result of an error.
+
+Time | Step | Operation | In memory "key" | Database "key" | Notes
+-----|------|-----------|-----------------|----------------|-------
+ t0 | 0 | start step | null | null | nothing stored yet
+ t1 | 0 | `put("key", "foo")` | "foo" | null | in-memory working map set
+ t2 | 0 | error happens | "foo" | null |
+ t3 | 0 | switch to undo | "foo" | "foo" | step state logged to database
+ t4 | 0 | start undo  | "foo" | "foo" | initial state of undo
+ t5 | 0 | `put("key", "xyz")` | "xyz" | "foo" | in-memory working map set
+ t6 | 0 | end step | "xyz" | "xyz" | step state logged to database
+
+As you can see from the table, the switch from do to undo records state in the Stairway
+database. Clearly it must; on server failure/recovery Stairway has to know that it is
+performing an undo. 
+
+Stairway makes the same guarantee about the state at the start of `undo()` as it does for the
+start of `do()`: the initial state will be the same even if there are server failures and
+recovery.
+
+### Working Map Best Practices <a name="workingmapbest"></a>
+
+A useful rule of thumb for the working map is: only use the working map to communicate
+data to later steps.
+
+Do not try to use the working map to communicate between the `do()`
+and the `undo()` of a step. While technically possible, it is difficult to make it
+reliable across server failure/recover.
+
+
+## Idempotent Steps <a name="idempotentsteps"></a>
+One of the features of Stairway is that it can resume flights after failure. Suppose a
+step in a flight is running and the server process is killed (e.g., the Kubernetes pod is
+deleted) the flight can be recovered. On recovery, the step that was running at the point
+of failure, is restarted.
+
+As a result, the `do()` and `undo()` methods of a step must be written to be
+idempotent; that is, they can be run multiple times and must have the same effect each
+time. Meeting that requirement often uses a different approach than writing a sequential program.
+
+### Database Transactions <a name="databasetransactions"></a>
+A step that consists of a database transaction is simple to write. The underlying database
+system provides the rollback of the transaction in the event of failure. However, it is
+still possible for a database transaction step to fail in the window between when the
+database transaction commits and when Stairway logs the completion of the step.
+
+You still have to make the step idempotent, so you have to handle that condition. Let's
+say you are inserting a new object into the database. This is best done by giving your
+target table a primary key column that is a UUID. Then you can run a sequence of steps
+something like this:
+ 1. Step that generates a random UUID and stores it in the working map; _OR_ hand a random
+ UUID into the flight as an input parameter. If this step is run more than once, it will
```

Comment:
here it's not clear you're not talking about the input parameter method and going back to the first approach.